client/src/main/java/org/opensearch/security/client/resources/ResourceSharingNodeClient.java

```diff
@@ -0,0 +1,246 @@
+
+        ListAccessibleResourcesRequest request = new ListAccessibleResourcesRequest.Builder().resourceIndex(resourceIndex).build();
+
+        client.execute(
+            ListAccessibleResourcesAction.INSTANCE,
+            request,
+            ActionListener.wrap(response -> listener.onResponse(response.getResources()), listener::onFailure)
+        );
+    }
+
+    /**
+     * Checks whether security or resource sharing is disabled and invokes failure on the listener if so.
+     *
+     * @param contextMessage  The context-specific message to log.
+     * @param listener The listener to notify with the exception.
+     * @return {@code true} if the feature is disabled, otherwise {@code false}.
+     */
+    private boolean handleIfDisabled(String contextMessage, ActionListener<?> listener) {
+        String featureDisabledMessage = securityOrFeatureDisabledMessage();
+        if (!featureDisabledMessage.isEmpty()) {
+            handleFeatureDisabled(featureDisabledMessage + " " + contextMessage, listener);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Checks whether security or resource sharing is disabled and responds with a default value if so.
+     *
+     * @param message        The context-specific message to log.
+     * @param listener       The listener to notify.
+     * @param defaultResponse The default response value to send.
+     * @param <T>            The type of the default response.
+     * @return {@code true} if the feature is disabled, otherwise {@code false}.
+     */
+    private <T> boolean handleIfDisabled(String message, ActionListener<T> listener, T defaultResponse) {
+        String fullMessage = securityOrFeatureDisabledMessage();
+        if (!fullMessage.isEmpty()) {
+            log.debug("{} {}", fullMessage, message);
+            listener.onResponse(defaultResponse);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Determines whether the security plugin or resource sharing feature is disabled.
+     *
+     * @return A non-empty message if disabled, otherwise an empty string.
+     */
+    private String securityOrFeatureDisabledMessage() {
+        boolean sharingEnabled = settings.getAsBoolean(
+            ConfigConstants.OPENSEARCH_RESOURCE_SHARING_ENABLED,
+            ConfigConstants.OPENSEARCH_RESOURCE_SHARING_ENABLED_DEFAULT
+        );
+
+        Settings securitySettings = settings.getByPrefix(ConfigConstants.SECURITY_SETTINGS_PREFIX);
+        boolean securityDisabled = securitySettings.isEmpty()
+            || this.settings.getAsBoolean(
+                ConfigConstants.OPENSEARCH_SECURITY_DISABLED,
+                ConfigConstants.OPENSEARCH_SECURITY_DISABLED_DEFAULT
+            );
+
+        if (securityDisabled) return "Security Plugin is disabled.";
+        if (!sharingEnabled) return "ShareableResource Access Control feature is disabled.";
+        return "";
+    }
```

Comment:
Nit: do we want to check here whether system index protections is enabled, and also return a message if so? Since the feature is dependent on that?